File: cpp/200_Number_of_Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                if (grid[i][j] == '1') {
                    count++;
                    BFS(grid, i, j);
                }
            }
        }
        return count;
    }

    void BFS(vector<vector<char>> &grid, int i, int j) {
        if (
            i < 0 
            || i >= grid.size() 
            || j < 0
            || j >= grid[i].size() 
            || grid[i][j] == '0'
        ) {
            return;
        }

        grid[i][j] = '0';
        BFS(grid, i + 1, j);
        BFS(grid, i - 1, j);
        BFS(grid, i, j + 1);
        BFS(grid, i, j - 1);
    }
};

File: php/0200/200_Number_of_Islands.php
<?php

/** class Solution */
class Solution
{

    /**
     * @param String[][] $grid
     * @return Integer
     */
    function numIslands($grid)
    {
        $count = 0;
        foreach ($grid as $keyRow => $arRow) {
            foreach ($arRow as $keyCell => $cell) {
                if ($grid[$keyRow][$keyCell] == '1') {
                    $count += 1;
                    $this->BFS($grid, $keyRow, $keyCell);
                }
            }
        }

        return $count;
    }

    /**
     * @param String[][] &$grid
     * @param Integer $keyRow
     * @param Integer $keyCell
     */
    protected function BFS(&$grid, $keyRow, $keyCell) 
    {
        if (
            $keyRow < 0
            || $keyRow >= count($grid)
            || $keyCell < 0
            || $keyCell >= count($grid[$keyRow])
            || $grid[$keyRow][$keyCell] == '0'
        ) {
            return;
        }

        $grid[$keyRow][$keyCell] = '0';
        
        $this->BFS($grid, $keyRow + 1, $keyCell);
        $this->BFS($grid, $keyRow - 1, $keyCell);
        $this->BFS($grid, $keyRow, $keyCell + 1);
        $this->BFS($grid, $keyRow, $keyCell - 1);
    }
}

